### ISSUE TITLE
SQL Server connector only allows globally unique index names

### ISSUE THREAD

**Issue by luisrudge** [prisma/prisma#6976]
After running `introspect`, prisma is throwing this error because of the index names. This might cause issues to keep evolving the model moving forward because the migrations would go crazy with the indexes 😬 

![image](https://user-images.githubusercontent.com/941075/117345917-7d398a00-ae7d-11eb-874c-39c8e209e370.png)

**Comment by janpio** [comment]
Can you possibly provide the SQL for this table (or tables, `Clients` and `Users`) so we can reproduce this manually? Would help us a lot.

**Comment by luisrudge** [comment]
@janpio sure!

```sql
SET ANSI_NULLS ON
GO
SET QUOTED_IDENTIFIER ON
GO
CREATE TABLE [dbo].[Clients](
	[Id] [int] IDENTITY(1,1) NOT NULL,
	[IsActive] [bit] NOT NULL,
	[Name] [nvarchar](100) NOT NULL,
	[Manager_UserID] [int] NOT NULL
) ON [PRIMARY]
GO
ALTER TABLE [dbo].[Clients] ADD  CONSTRAINT [PK_dbo.Clients] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
CREATE NONCLUSTERED INDEX [IX_Manager_UserID] ON [dbo].[Clients]
(
	[Manager_UserID] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, DROP_EXISTING = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
ALTER TABLE [dbo].[Clients]  WITH NOCHECK ADD  CONSTRAINT [FK_dbo.Clients_dbo.Users_Manager_UserID] FOREIGN KEY([Manager_UserID])
REFERENCES [dbo].[Users] ([UserID])
GO
ALTER TABLE [dbo].[Clients] CHECK CONSTRAINT [FK_dbo.Clients_dbo.Users_Manager_UserID]
GO
```

```sql
SET ANSI_NULLS ON
GO
SET QUOTED_IDENTIFIER ON
GO
CREATE TABLE [dbo].[Users](
	[UserID] [int] IDENTITY(1,1) NOT NULL,
	[AccessGuid] [uniqueidentifier] NULL,
	[Address] [nvarchar](200) NULL,
	[City] [nvarchar](100) NULL,
	[CreationDate] [datetime] NOT NULL,
	[HasAdminAccess] [bit] NOT NULL,
	[IsActive] [bit] NOT NULL,
	[LastLogin] [datetime] NULL,
	[Login] [nvarchar](100) NOT NULL,
	[Name] [nvarchar](200) NOT NULL,
	[Password] [nvarchar](100) NOT NULL,
	[Phone] [nvarchar](50) NULL,
	[PostalCode] [nvarchar](15) NULL,
	[State] [nvarchar](100) NULL,
	[ManagerID] [int] NULL,
	[IsRestricted] [bit] NOT NULL,
	[ExternalId] [nvarchar](max) NOT NULL
) ON [PRIMARY] TEXTIMAGE_ON [PRIMARY]
GO
ALTER TABLE [dbo].[Users] ADD  CONSTRAINT [PK_dbo.Users] PRIMARY KEY CLUSTERED 
(
	[UserID] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
CREATE UNIQUE NONCLUSTERED INDEX [IX_AccessGuid] ON [dbo].[Users]
(
	[AccessGuid] ASC
)
WHERE ([AccessGuid] IS NOT NULL)
WITH (STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, DROP_EXISTING = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
CREATE NONCLUSTERED INDEX [IX_ManagerID] ON [dbo].[Users]
(
	[ManagerID] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, DROP_EXISTING = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
SET ANSI_PADDING ON
GO
CREATE UNIQUE NONCLUSTERED INDEX [IX_USER_LOGIN] ON [dbo].[Users]
(
	[Login] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, DROP_EXISTING = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
ALTER TABLE [dbo].[Users] ADD  DEFAULT ((0)) FOR [HasAdminAccess]
GO
ALTER TABLE [dbo].[Users] ADD  DEFAULT ((0)) FOR [IsRestricted]
GO
ALTER TABLE [dbo].[Users] ADD  DEFAULT ('') FOR [ExternalId]
GO
ALTER TABLE [dbo].[Users]  WITH NOCHECK ADD  CONSTRAINT [FK_dbo.Users_dbo.Users_ManagerID] FOREIGN KEY([ManagerID])
REFERENCES [dbo].[Users] ([UserID])
GO
ALTER TABLE [dbo].[Users] CHECK CONSTRAINT [FK_dbo.Users_dbo.Users_ManagerID]
GO

```

**Comment by luisrudge** [comment]
sorry, this is generated by sql server itself, so it has lots of extra commands 😬

**Comment by janpio** [comment]
Now I only notice this in the image: _Is_ `IX_Manager_UserID` actually defined multiple times as `@@index` in your file?

(Maybe you can also share the part of the schema you took a screenshot off - this might just be a visual problem even)

**Comment by luisrudge** [comment]
Oh, I probably forgot another table (the one that uses the same index).

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["microsoftSqlServer"]
}

datasource db {
  provider = "sqlserver"
  url      = env("DATABASE_URL")
}

model Classes {
  Id                  Int       @id @default(autoincrement())
  End                 DateTime? @db.DateTime
  IsActive            Boolean
  Name                String    @db.NVarChar(100)
  Start               DateTime  @db.DateTime
  Client_Id           Int
  Manager_UserID      Int
  Clients             Clients   @relation(fields: [Client_Id], references: [Id])
  Users               Users     @relation(fields: [Manager_UserID], references: [UserID])

  @@index([Client_Id], name: "IX_Client_Id")
  @@index([Manager_UserID], name: "IX_Manager_UserID")
}

model Clients {
  Id             Int       @id @default(autoincrement())
  IsActive       Boolean
  Name           String    @db.NVarChar(100)
  Manager_UserID Int
  Users          Users     @relation(fields: [Manager_UserID], references: [UserID])
  Classes        Classes[]

  @@index([Manager_UserID], name: "IX_Manager_UserID")
}

model Users {
  UserID         Int       @id @default(autoincrement())
  AccessGuid     String?   @db.UniqueIdentifier
  Address        String?   @db.NVarChar(200)
  City           String?   @db.NVarChar(100)
  CreationDate   DateTime  @db.DateTime
  HasAdminAccess Boolean   @default(false)
  IsActive       Boolean
  LastLogin      DateTime? @db.DateTime
  Login          String    @unique @db.NVarChar(100)
  Name           String    @db.NVarChar(200)
  Password       String    @db.NVarChar(100)
  Phone          String?   @db.NVarChar(50)
  PostalCode     String?   @db.NVarChar(15)
  State          String?   @db.NVarChar(100)
  ManagerID      Int?
  IsRestricted   Boolean   @default(false)
  ExternalId     String    @default("") @db.NVarChar(Max)
  Users          Users?    @relation("UsersToUsers_ManagerID", fields: [ManagerID], references: [UserID])
  Clients        Clients[]

  @@index([ManagerID], name: "IX_ManagerID")
}
```

```sql
SET ANSI_NULLS ON
GO
SET QUOTED_IDENTIFIER ON
GO
CREATE TABLE [dbo].[Classes](
	[Id] [int] IDENTITY(1,1) NOT NULL,
	[End] [datetime] NULL,
	[IsActive] [bit] NOT NULL,
	[Name] [nvarchar](100) NOT NULL,
	[Start] [datetime] NOT NULL,
	[Client_Id] [int] NOT NULL,
	[Manager_UserID] [int] NOT NULL
) ON [PRIMARY]
GO
ALTER TABLE [dbo].[Classes] ADD  CONSTRAINT [PK_dbo.Classes] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
CREATE NONCLUSTERED INDEX [IX_Client_Id] ON [dbo].[Classes]
(
	[Client_Id] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, DROP_EXISTING = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
CREATE NONCLUSTERED INDEX [IX_Manager_UserID] ON [dbo].[Classes]
(
	[Manager_UserID] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, DROP_EXISTING = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO
ALTER TABLE [dbo].[Classes]  WITH NOCHECK ADD  CONSTRAINT [FK_dbo.Classes_dbo.Clients_Client_Id] FOREIGN KEY([Client_Id])
REFERENCES [dbo].[Clients] ([Id])
ON DELETE CASCADE
GO
ALTER TABLE [dbo].[Classes] CHECK CONSTRAINT [FK_dbo.Classes_dbo.Clients_Client_Id]
GO
ALTER TABLE [dbo].[Classes]  WITH NOCHECK ADD  CONSTRAINT [FK_dbo.Classes_dbo.Users_Manager_UserID] FOREIGN KEY([Manager_UserID])
REFERENCES [dbo].[Users] ([UserID])
GO
ALTER TABLE [dbo].[Classes] CHECK CONSTRAINT [FK_dbo.Classes_dbo.Users_Manager_UserID]
GO

```

**Comment by janpio** [comment]
Right, the error message complains about "globally unique". That does not seem to be true for SQL Server, while it is probably true for some other database.

Code for this: https://github.com/prisma/prisma-engines/blob/789a706435edd011021ae544bad8e2bf582efd9f/libs/datamodel/core/src/transform/ast_to_dml/validate.rs#L216-L252
So `supports_multiple_indexes_with_same_name()` is wrong for the SQLServer connector probably.

**Comment by luisrudge** [comment]
I'm glad you found out!

**Comment by janpio** [comment]
And it should be fixed :D

**Comment by luisrudge** [comment]
amazing! When can we expect this to be released?

**Comment by janpio** [comment]
Very soon. When https://github.com/prisma/prisma/pull/6971 is merged actually, that will create a new `dev` version of Prisma to test this. And in a proper release it will be included on May 18 in 1.5 weeks.
